Support token revocation

michaelsilver opened the pull request:
Adds endpoint `ctrl/b/:backend/:user/revoke-token` which deletes any tokens matching the given backend and user combination.

This permits revoking connection tokens based on the (username, backend ID) pair, while leaving tokens without a username as non-revocable.

Fixes https://github.com/drifting-in-space/plane/issues/587

Test plan
=========

Create a token for the user `user-123`:
```bash
$ curl -X "POST" "http://localhost:8080/ctrl/connect" \
     -H 'Content-Type: application/json; charset=utf-8' \
     -d $'{
  "spawn_config": {
    "lifetime_limit_seconds": 3600,
    "max_idle_seconds": 60,
    "executable": {
      "image": "ghcr.io/drifting-in-space/demo-image-drop-four"
    }
  },
  "key": {
    "name": "room-abcde",
    "namespace": "games",
    "tag": "drop-four"
  },
  "auth": {
    "even lists and numbers": [
      "like",
      3.14,
      "are",
      "fine"
    ],
    "any arbitrary JSON object": "can go here",
    "Plane does not read it": "it is just passed through to your backend"
  },
  "user": "user-123"
}'
```

Delete `user-123`'s token (where in this instance, `vnb8zz77dm1mfi` is the `backend_id`):
```bash
$ curl -X "POST" "http://localhost:8080/ctrl/b/vnb8zz77dm1mfi/user-123/revoke-token"
```

Verify that the token gets deleted:
```
$ ./dev/psql.sh

postgres=# select token, username from token;
                    token                    | username
---------------------------------------------+----------
 014qfddf2INksUB4W2TCpucTrn4haUdztQ10SnGoJz4 |
 u6kTml5AGgaLXHWDeTz5__2UR6bb26uldPNa-MTTf9A |
 osEGZaNuWa5xiyb7UnUzk1f7eLwc9I2PfuqBim7iYos |
 34bj1njNYUy2UBQeSRlFq1czQT05MtD4UnLJwURE4bI |
 jywfnWxAlCJQRhk4KR_yzvBWyAOBmVmEodSW6B7mVaA |
 oDn_3bLf_myCoPqafXEBXkKGUJW14jhibAmU8PU4ac8 |
 KsuyJeGWofuEYvTfwbdeAK6mtlNj9oEjs2RndPZf_Po | user-123
(7 rows)

postgres=# select token, username from token;
                    token                    | username
---------------------------------------------+----------
 014qfddf2INksUB4W2TCpucTrn4haUdztQ10SnGoJz4 |
 u6kTml5AGgaLXHWDeTz5__2UR6bb26uldPNa-MTTf9A |
 osEGZaNuWa5xiyb7UnUzk1f7eLwc9I2PfuqBim7iYos |
 34bj1njNYUy2UBQeSRlFq1czQT05MtD4UnLJwURE4bI |
 jywfnWxAlCJQRhk4KR_yzvBWyAOBmVmEodSW6B7mVaA |
 oDn_3bLf_myCoPqafXEBXkKGUJW14jhibAmU8PU4ac8 |
(6 rows)

michaelsilver commented:
Other discussion points:
* whether we want this functionality implemented in the client code (I did, figuring it’d help with testing in the future — happy to remove though!)
* whether we want this functionality implemented in the CLI (I didn’t, given there was no other mention of `user` in the CLI currently — would be easy to add if we wanted!)

paulgb commented:
Thanks for this! The approach is sound.

One additional step we need to take is to have the proxies made aware when a token is revoked, because they currently cache the token->backend mapping. But I think this will require a larger re-thinking of how data are synchronized between the controller and proxies, so I'm ok with building the functionality into the API this way as long as we have a comment that mentions that limitation.

Eventually we could even have it propagate to the proxies all the way so that it even interrupts existing connections.

paulgb commented:
Sorry, I hit "update branch" by mistake, you will need to pull again before you can push.

michaelsilver commented:
Back to work on Plane, sorry for the delay here! 

Added a comment, cleaned up the commits/rebased on main + double checked everything works as expected.

No code changes so should be an easy stamp

paulgb reviewed (commented):
Looks good. Left a couple of comments.

One request we've had is for revoking to work on a (username, key) pair rather than (username, backend), since that mirrors how they are assigned. I think that could be part of a bigger refactor that scopes connection tokens to keys rather than backends, though, so I'm fine deferring that and keeping this functionality for now.

michaelsilver commented on plane/src/client/mod.rs:
```diff
@@ -138,6 +137,19 @@ impl PlaneClient {
+    pub async fn revoke_token(
+        &self,
+        backend_id: &BackendName,
+        username: &str,
```
Ah good catch, went with the second option to move `backend_id` and `user` into a JSON payload because it feels more consistent given we handle `user` in JSON in `connect`. 

I took the opportunity to clean up the code a bit and leverage typing more so we don't have to redo a bunch of function signatures should we want to change the arguments to the API (like allowing folks to use `key` etc.). 

The API is now `/ctrl/revoke` and renamed a bunch of functions accordingly. Lmk if you prefer a different naming like `revoke-token` though.

michaelsilver commented on plane/src/database/connect.rs:
```diff
@@ -191,6 +191,20 @@ async fn create_token(
+    sqlx::query!(
+        r#"
+        DELETE FROM token
+        WHERE backend_id = $1 AND username = $2
```
Fixed and noted for the future!

michaelsilver commented:
@paulgb Thanks -- ready for another quick review! Added the changes as separate commits in case it's easier to see the diffs, but I can rebase/squash those before we merge.

paulgb commented:
LGTM, thanks for this!
